File: backend/core/file_types.py
# --- Filename or extension-based patterns ---
# Aligned with UI AdvancedFiltersPanel.jsx fileTypeGroups while keeping existing types for compatibility
FILE_TYPE_PATTERNS = {
    # UI: Documents
    "Text Documents": [".txt", ".doc", ".docx", ".pdf", ".rtf", ".odt"],
    "Spreadsheets": [".xls", ".xlsx", ".csv", ".ods"],
    "Presentations": [".ppt", ".pptx", ".odp", ".key"],

    # UI: Media
    "Images": [".jpg", ".jpeg", ".png", ".gif", ".svg", ".bmp", ".webp", ".ico"],
    "Audio": [".mp3", ".wav", ".flac", ".aac", ".ogg", ".m4a", ".wma"],
    "Video": [".mp4", ".mov", ".avi", ".mkv", ".wmv", ".flv", ".webm"],

    # UI: Development
    "Code": [".js", ".ts", ".jsx", ".tsx",".py", ".java", ".cpp", ".c", ".rb", ".go", ".rs", ".swift"],
    "Web": [".html", ".css", ".jsx", ".tsx", ".vue", ".scss", ".sass", ".less"],
    "Config": [".json", ".yaml", ".yml", ".xml", ".toml", ".ini", ".conf"],
    "Scripts": [".sh", ".bash", ".zsh", ".ps1", ".bat", ".cmd", ".command"],

    # UI: Archives
    "Compressed": [".zip", ".rar", ".tar", ".gz", ".7z", ".bz2", ".xz"],

    # Existing semantic buckets (kept for compatibility)
    "Models": ["Model", "Entity"],
    "Views": ["View", ".xib", ".storyboard", "UI", "Layout"],
    "ViewModels": ["ViewModel", "Binder"],
    "Services": ["Service", "Manager", "Helper", "Fetcher", "Syncer"],
    "Controllers": ["Controller", "Coordinator", "Router"],
    "Utilities": ["Utils", "Utility", "Common", "Shared"],
    "Tests": ["Test", "Spec", "Mock"],
    "MainApp": ["AppDelegate", "SceneDelegate", "Main", "Entry", "main.swift", "main.py"],

    # Existing languages (kept for compatibility)
    "Swift": [".swift"],
    "Python": [".py"],
    "JavaScript": [".js"],
    "TypeScript": [".ts"],
    "HTML": [".html", ".htm"],
    "CSS": [".css"],
    "C/C++": [".c", ".cpp", ".h", ".hpp"],
    "Java": [".java"],
    "Kotlin": [".kt"],
    "Dart": [".dart"],
    "Rust": [".rs"],
    "Go": [".go"],

    # Other/general (kept)
    "Documents": [".docx", ".pdf", ".txt", ".rtf", ".md"],
    "Media": [".mp4", ".mov", ".avi", ".mp3", ".wav", ".flac", ".mkv"],
    "Executables": [".exe", ".dmg", ".app", ".apk", ".msi"],
    "Archives": [".zip", ".tar", ".gz", ".rar", ".7z", ".iso"],
    "Design": [".psd", ".ai", ".fig", ".sketch"],
    "Security": [".pem", ".crt", ".key", ".pfx"],
    "Assets": ["asset", "icon", "logo", "background"]
}

# --- Content-based markers for deep scan ---
CONTENT_MARKERS = {
    # UI: Documents
    "Text Documents": ["Title:", "Author:", "Table of Contents", "Lorem ipsum"],
    "Spreadsheets": ["Sheet1", "SUM(", "VLOOKUP", "PivotTable"],
    "Presentations": ["Slide", "PowerPoint", "Keynote", "Agenda"],

    # UI: Media
    "Images": ["Exif", "RGB", "Alpha Channel", "Layer"],
    "Audio": ["codec", "bitrate", "duration", "sample rate"],
    "Video": ["codec", "frame rate", "bitrate", "duration"],

    # UI: Development
    "Code": ["function", "class", "def ", "import ", "package ", "public class"],
    "Web": ["<!DOCTYPE html>", "<html", "className", "export default", "Vue", "style {", "scss", "less"],
    "Config": ["{", ":", "version", "config", "root", "schema"],
    "Scripts": ["#!/usr/bin", "bash", "powershell", "echo", "set ", "Get-"],

    # UI: Archives
    "Compressed": ["PK", "7z", "XZ", "BZh", "ustar"],

    # Existing architecture
    "Models": ["class Model", "struct Model", "typealias Model", "extension Model", "Codable", "Entity"],
    "Views": ["body: some View", "UIView", "render()", "drawRect", "SwiftUI", "class View"],
    "ViewModels": ["class ViewModel", "ObservableObject", "@Published", "bind()", "updateView"],
    "Services": ["class Service", "func fetch", "func sync", "handleRequest", "performTask"],
    "Controllers": ["class Controller", "class Coordinator", "func navigate", "func handle", "present", "dismiss"],
    "Utilities": ["extension String", "extension Array", "sharedInstance", "singleton"],
    "Tests": ["XCTest", "unittest", "describe(", "it(", "assert", "mock", "stub"],
    "MainApp": ["@main", "UIApplication", "AppDelegate", "SceneDelegate", "main()"],

    # Existing language-specific
    "Swift": ["import SwiftUI", "class", "struct", "extension", "func"],
    "Python": ["def ", "import ", "class ", "if __name__ == '__main__'"],
    "JavaScript": ["function(", "const ", "let ", "import ", "export "],
    "TypeScript": ["interface ", "type ", "class ", "import ", "export "],
    "HTML": ["<!DOCTYPE html>", "<html>", "<head>", "<body>"],
    "CSS": ["color:", "font-family:", "margin:", "padding:"],
    "C/C++": ["#include", "int main(", "void ", "class ", "struct "],
    "Java": ["public class", "System.out.println", "import java."],
    "Kotlin": ["fun ", "val ", "var ", "class ", "object "],
    "Dart": ["void main()", "import ", "class ", "Widget"],
    "Rust": ["fn main()", "use ", "struct ", "impl "],
    "Go": ["package main", "func main()", "import "],

    # Other/general
    "Documents": ["Title:", "Author:", "Table of Contents", "Lorem ipsum"],
    "Media": ["codec", "frame rate", "bitrate", "duration"],
    "Executables": ["MZ", "PK", "Mach-O", "ELF"],
    "Archives": ["PK", "compressed", "archive", "tarball"],
    "Design": ["Adobe", "Sketch", "Figma", "Artboard"],
    "Security": ["BEGIN CERTIFICATE", "BEGIN PRIVATE KEY", "SSL", "TLS"],
    "Assets": ["imageNamed", "loadImage", "UIImage", "NSImage"]
}

# --- UI color mapping for semantic types ---
TYPE_COLORS = {
    # UI-aligned
    "Text Documents": "#e3f2fd",
    "Spreadsheets": "#fff3e0",
    "Presentations": "#ede7f6",
    "Images": "#fce4ec",
    "Audio": "#e1f5fe",
    "Video": "#f3e5f5",
    "Code": "#d1c4e9",
    "Web": "#ffccbc",
    "Config": "#f9fbe7",
    "Scripts": "#f3e5f5",
    "Compressed": "#f0f4c3",

    # Existing architecture
    "Models": "#e0f7fa",
    "Views": "#fce4ec",
    "ViewModels": "#f3e5f5",
    "Services": "#e8f5e9",
    "Controllers": "#fff3e0",
    "Utilities": "#f5f5f5",
    "Tests": "#ede7f6",
    "MainApp": "#c8e6c9",

    # Existing languages
    "Swift": "#d1c4e9",
    "Python": "#ffe0b2",
    "JavaScript": "#fff176",
    "TypeScript": "#b3e5fc",
    "HTML": "#ffccbc",
    "CSS": "#cfd8dc",
    "C/C++": "#b0bec5",
    "Java": "#ffab91",
    "Kotlin": "#ce93d8",
    "Dart": "#80deea",
    "Rust": "#bcaaa4",
    "Go": "#aed581",

    # Other/general
    "Documents": "#e8f5e9",
    "Media": "#e1f5fe",
    "Executables": "#fbe9e7",
    "Archives": "#f0f4c3",
    "Design": "#f8bbd0",
    "Security": "#d7ccc8",
    "Assets": "#fffde7",

    # Fallbacks
    "Unclassified": "#ffffff",
    "Error": "#ffebee"
}

# --- UI grouping for semantic filters ---
TYPE_GROUPS = {
    # Keep original programming groups
    "Programming Architecture": [
        "Models", "Views", "ViewModels", "Services", "Controllers",
        "Utilities", "Tests", "MainApp"
    ],
    "Programming Languages": [
        "Swift", "Python", "JavaScript", "TypeScript", "HTML", "CSS",
        "C/C++", "Java", "Kotlin", "Dart", "Rust", "Go"
    ],

    # UI-aligned groups
    "Documents": ["Text Documents", "Spreadsheets", "Presentations"],
    "Media": ["Images", "Audio", "Video"],
    "Development": ["Code", "Web", "Config", "Scripts"],
    "System": ["Compressed", "Executables", "Fonts", "Database"]
}

def get_file_label_from_name(filename: str) -> str:
    """
    Returns a short label (extension) from a filename.
    """
    if "." not in filename or filename.startswith("."):
        return "FILE"
    return filename.split(".")[-1].upper()

def get_file_category(filename: str) -> str:
    """
    Determines the semantic category of a file based on its extension.
    Used for the 'Organize by Type' copy feature.
    """
    lower_name = filename.lower()
    
    # Check strict extension match first
    for category, extensions in FILE_TYPE_PATTERNS.items():
        for ext in extensions:
            if lower_name.endswith(ext):
                return category
    
    return None

File: src/utils/fileTypes.js
import {
  FileText,
  Image,
  Music,
  Video,
  Code,
  Archive,
  File,
} from "lucide-react";

/**
 * File Type Definitions for Frontend
 * Aligned with backend/core/file_types.py
 */
export const fileTypeGroups = {
  Documents: [
    {
      name: "Text Documents",
      icon: FileText,
      types: [
        ".txt",
        ".doc",
        ".docx",
        ".pdf",
        ".rtf",
        ".odt",
        ".md",
        ".tex",
        ".wpd",
      ],
    },
    {
      name: "Spreadsheets",
      icon: FileText,
      types: [".xls", ".xlsx", ".csv", ".ods", ".tsv"],
    },
    {
      name: "Presentations",
      icon: FileText,
      types: [".ppt", ".pptx", ".odp", ".key"],
    },
  ],
  Media: [
    {
      name: "Images",
      icon: Image,
      types: [
        // Standard
        ".jpg",
        ".jpeg",
        ".png",
        ".gif",
        ".svg",
        ".bmp",
        ".webp",
        ".ico",
        ".tiff",
        ".tif",
        // High Efficiency
        ".heic",
        ".heif",
        ".avif",
        // RAW Formats (Professional)
        ".cr2",
        ".cr3",
        ".crw", // Canon
        ".nef",
        ".nrw", // Nikon
        ".arw",
        ".srf",
        ".sr2", // Sony
        ".orf", // Olympus
        ".rw2", // Panasonic
        ".raf", // Fujifilm
        ".dng", // Adobe / Generic
        ".pef", // Pentax
        ".x3f", // Sigma
        ".3fr",
        ".fff", // Hasselblad
        ".iiq", // Phase One
      ],
    },
    {
      name: "Audio",
      icon: Music,
      types: [
        ".mp3",
        ".wav",
        ".flac",
        ".aac",
        ".ogg",
        ".m4a",
        ".wma",
        ".aiff",
        ".alac",
      ],
    },
    {
      name: "Video",
      icon: Video,
      types: [
        ".mp4",
        ".mov",
        ".avi",
        ".mkv",
        ".wmv",
        ".flv",
        ".webm",
        ".m4v",
        ".mpg",
        ".mpeg",
        ".3gp",
        ".mts",
        ".m2ts",
      ],
    },
  ],
  Development: [
    {
      name: "Code",
      icon: Code,
      types: [
        ".js",
        ".ts",
        ".jsx",
        ".tsx",
        ".py",
        ".java",
        ".cpp",
        ".c",
        ".rb",
        ".go",
        ".rs",
        ".swift",
        ".php",
        ".sql",
      ],
    },
    {
      name: "Web",
      icon: Code,
      types: [".html", ".css", ".scss", ".sass", ".less", ".vue"],
    },
    {
      name: "Config",
      icon: FileText,
      types: [
        ".json",
        ".yaml",
        ".yml",
        ".xml",
        ".toml",
        ".ini",
        ".conf",
        ".env",
      ],
    },
    {
      name: "Scripts",
      icon: File,
      types: [
        ".sh",
        ".bash",
        ".zsh",
        ".ps1",
        ".bat",
        ".cmd",
        ".command",
        ".lua",
        ".pl",
      ],
    },
  ],
  Archives: [
    {
      name: "Compressed",
      icon: Archive,
      types: [
        ".zip",
        ".rar",
        ".tar",
        ".gz",
        ".7z",
        ".bz2",
        ".xz",
        ".iso",
        ".dmg",
      ],
    },
  ],
};

// Extension → human-friendly label
const extensionToLabel = {
  // Documents
  ".doc": "Word",
  ".docx": "Word",
  ".odt": "OpenDocument Text",
  ".rtf": "Rich Text",
  ".pdf": "PDF",
  ".txt": "Plain Text",
  ".md": "Markdown",
  ".tex": "LaTeX",
  ".wpd": "WordPerfect",
  ".csv": "CSV",
  ".xls": "Excel",
  ".xlsx": "Excel",
  ".ods": "OpenDocument Spreadsheet",
  ".tsv": "Tab Separated",
  ".ppt": "PowerPoint",
  ".pptx": "PowerPoint",
  ".odp": "OpenDocument Presentation",
  ".key": "Keynote",

  // Images (Standard)
  ".jpg": "JPEG",
  ".jpeg": "JPEG",
  ".png": "PNG",
  ".gif": "GIF",
  ".svg": "SVG",
  ".bmp": "BMP",
  ".webp": "WebP",
  ".ico": "ICO",
  ".tiff": "TIFF",
  ".tif": "TIFF",

  // Images (High Efficiency)
  ".heic": "HEIC Image",
  ".heif": "HEIF Image",
  ".avif": "AVIF Image",

  // Images (RAW)
  ".cr2": "Canon RAW",
  ".cr3": "Canon RAW",
  ".crw": "Canon RAW",
  ".nef": "Nikon RAW",
  ".nrw": "Nikon RAW",
  ".arw": "Sony RAW",
  ".srf": "Sony RAW",
  ".sr2": "Sony RAW",
  ".orf": "Olympus RAW",
  ".rw2": "Panasonic RAW",
  ".raf": "Fujifilm RAW",
  ".dng": "Digital Negative",
  ".pef": "Pentax RAW",
  ".x3f": "Sigma RAW",
  ".3fr": "Hasselblad RAW",
  ".fff": "Hasselblad RAW",
  ".iiq": "Phase One RAW",

  // Audio
  ".mp3": "MP3 Audio",
  ".wav": "WAV Audio",
  ".flac": "FLAC Audio",
  ".aac": "AAC Audio",
  ".ogg": "Ogg Audio",
  ".m4a": "M4A Audio",
  ".wma": "WMA Audio",
  ".aiff": "AIFF Audio",
  ".alac": "Apple Lossless",

  // Video
  ".mp4": "MP4 Video",
  ".mov": "QuickTime Video",
  ".avi": "AVI Video",
  ".mkv": "Matroska Video",
  ".wmv": "Windows Media",
  ".flv": "Flash Video",
  ".webm": "WebM Video",
  ".m4v": "M4V Video",
  ".mpg": "MPEG Video",
  ".mpeg": "MPEG Video",
  ".3gp": "3GP Video",
  ".mts": "AVCHD Video",
  ".m2ts": "Blu-ray Video",

  // Code / Web
  ".js": "JavaScript",
  ".jsx": "React JSX",
  ".ts": "TypeScript",
  ".tsx": "React TSX",
  ".py": "Python",
  ".java": "Java",
  ".cpp": "C++",
  ".c": "C",
  ".rb": "Ruby",
  ".go": "Go",
  ".rs": "Rust",
  ".swift": "Swift",
  ".php": "PHP",
  ".sql": "SQL",
  ".html": "HTML",
  ".css": "CSS",
  ".scss": "Sass (SCSS)",
  ".sass": "Sass",
  ".less": "Less",
  ".vue": "Vue Component",

  // Config
  ".json": "JSON",
  ".yaml": "YAML",
  ".yml": "YAML",
  ".xml": "XML",
  ".toml": "TOML",
  ".ini": "INI",
  ".conf": "Config",
  ".env": "Environment",

  // Scripts
  ".sh": "Shell Script",
  ".bash": "Bash Script",
  ".zsh": "Zsh Script",
  ".ps1": "PowerShell Script",
  ".bat": "Batch File",
  ".cmd": "Command Script",
  ".lua": "Lua Script",
  ".pl": "Perl Script",

  // Archives
  ".zip": "ZIP Archive",
  ".rar": "RAR Archive",
  ".tar": "TAR Archive",
  ".gz": "GZIP Archive",
  ".7z": "7-Zip Archive",
  ".bz2": "BZip2 Archive",
  ".xz": "XZ Archive",
  ".iso": "Disc Image",
  ".dmg": "Disk Image",
};

export const getExtension = (nameOrPath = "") => {
  const base = String(nameOrPath).split("/").pop() || "";
  const dot = base.lastIndexOf(".");
  if (dot < 0) return "";
  return base.slice(dot).toLowerCase();
};

export const getFileLabelFromName = (nameOrPath = "") => {
  const ext = getExtension(nameOrPath);
  if (!ext) return "Unknown";
  return extensionToLabel[ext] || ext.replace(".", "").toUpperCase();
};
